tests/TestCase/Cache/Engine/FileEngineTest.php

```diff
@@ -193,6 +207,34 @@ public function testSerialize()
     }
 
     /**
+     * Verify across PHP versions that caching of ResultSet objects works as expected.
+     *
+     * @see https://github.com/cakephp/cakephp/issues/10111
+     * @return void
+     */
+    public function testResultSetCache()
+    {
+        Configure::write('App.namespace', 'TestApp');
+        $this->loadFixtures('Articles');
+
+        /* @var ArticlesTable $articles */
+        $articles = TableRegistry::get('Articles');
+        $query = $articles->find()->all();
+        $this->assertInstanceOf(ResultSet::class, $query);
+        $this->assertSame(3, count($query->toArray()));
+
+        $this->_configCache(['serialize' => true]);
+        $write = Cache::write('articles', $query, 'file_test');
+        $this->assertTrue($write);
+
+        $read = Cache::read('articles', 'file_test');
+        $this->assertInstanceOf(ResultSet::class, $read);
+        $this->assertSame(3, count($read->toArray()));
+
+        TableRegistry::clear();
```

Comment:
Couldn't this test be a test for `ResultSet` being serialized and unserialized as that is where the actual issue was.